tests/test_tensordata.py

```diff
@@ -19,8 +19,121 @@ def almost_equal(match_tensor: TensorData, torch_tensor: torch.Tensor) -> bool:
 
 
+def same_references(match_tensor1: TensorData, match_tensor2: TensorData):
+    # check if they are the same object, which they should to match pytorch functionality
+    for e1, e2 in zip(match_tensor1._data, match_tensor2._data):
+        if e1 is not e2:
+            return False
+    return True
+
+
 # Customize with https://medium.com/@lucpham/how-to-customize-unittest-in-python-d4dfb83f1dba
 class TestTensorDataTest(unittest.TestCase):
+    def test_binary_operations(self):
+        torch_tensor_low_dim_1 = torch.ones(2, 2)
+        torch_tensor_low_dim_2 = torch.ones(2, 2)
+        torch_tensor_singleton = torch.ones(1)[0]
+        torch_tensor_high_dim = torch.ones(1, 3, 2, 1)
+
+        match_tensor_low_dim_1 = TensorData(2, 2)
+        match_tensor_low_dim_1.ones_()
+
+        match_tensor_low_dim_2 = TensorData(2, 2)
+        match_tensor_low_dim_2.ones_()
+
+        match_tensor_singleton = TensorData(value=1.0)
+
+        match_tensor_high_dim = TensorData(1, 3, 2, 1)
+        match_tensor_high_dim.ones_()
+
+        # low low
+        self.assertTrue(
+            almost_equal(
+                match_tensor_low_dim_1 + match_tensor_low_dim_2,
+                torch_tensor_low_dim_1 + torch_tensor_low_dim_2,
+            )
+        )
+
+        # singleton low
+        self.assertTrue(
+            almost_equal(
+                match_tensor_singleton + match_tensor_low_dim_2,
+                torch_tensor_singleton + torch_tensor_low_dim_2,
+            )
+        )
+
+        # singleton singleton
+        match_singleton_add = TensorData(value=1.0) + TensorData(value=1.0)
+        self.assertEqual(match_singleton_add.shape, ())
+        self.assertEqual(match_singleton_add._data, None)
+        self.assertEqual(match_singleton_add._item, 2)
+
+        # low high
+        self.assertTrue(
+            almost_equal(
+                match_tensor_low_dim_2 + match_tensor_high_dim,
+                torch_tensor_low_dim_2 + torch_tensor_high_dim,
+            )
+        )
+
+    def test_transpose(self):
+        # make torch tensor
+        torch_tensor = torch.arange(9).reshape(3, 1, 3)
+        # make corresponding match tensor
+        match_tensor = TensorData(3, 1, 3)
+        match_tensor._data = [TensorData(value=i) for i in range(9)]
+
+        self.assertTrue(almost_equal(match_tensor.T(), torch_tensor.T))
+
+    def test_permute(self):
+        # make torch tensor
+        torch_tensor = torch.arange(9).reshape(3, 1, 3)
+        # make corresponding match tensor
+        match_tensor = TensorData(3, 1, 3)
+        match_tensor._data = [TensorData(value=i) for i in range(9)]
+
+        self.assertTrue(
+            almost_equal(match_tensor.permute(2, 0, 1), torch_tensor.permute(2, 0, 1))
+        )
+
+    def test_reshape_failure(self):
+        match_tensor = TensorData(2, 3, 4)
+        self.assertRaises(RuntimeError, lambda: match_tensor.reshape(5, 5, 5))
+
+    def test_reshape(self):
```

Comment:
Should we have any tests checking if the values are in the correct order. For example, comparing our output to Pytorch?